src/coreclr/jit/morph.cpp

```diff
@@ -12096,6 +12081,23 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 op1  = tree->gtGetOp1();
                 op2  = tree->gtGetOp2();
             }
+
+            // Optimizes (X & 1) == 1 to (X & 1)
+            // GTF_RELOP_JMP_USED is used to make sure the optimization is used for return statements only.
+            if (tree->gtGetOp2()->IsIntegralConst(1) && tree->gtGetOp1()->OperIs(GT_AND) &&
+                !(tree->gtFlags & GTF_RELOP_JMP_USED))
+            {
+                GenTree* op1 = tree->gtGetOp1();
+
+                if (op1->gtGetOp2()->IsIntegralConst(1))
+                {
+                    DEBUG_DESTROY_NODE(tree->gtGetOp2());
+                    DEBUG_DESTROY_NODE(tree);
+
+                    return op1;
+                }
+            }
```

Comment:
(In post-order) ...and inside `fgOptimizeEqualityComparisonWithConst`.

(Sorry for not being complete in my first comment :( )

Reply:
I'm happy to hear any corrections :) feel free

Reply:
`(In post-order)` Do you mean the order the nodes are being destroyed?

Reply:
`fgOptimizeEqualityComparisonWithConst` is supposed to return a comparison node, meanwhile we have to get rid of the equality node. Are you sure the optimization should be placed to there?
https://github.com/dotnet/runtime/blob/99d82c20269d75e69a60ef4d544ca9d1efd93cb7/src/coreclr/jit/morph.cpp#L12085-L12086

Reply:
> Are you sure the optimization should be placed to there?

Yeah, this is an artificial restriction and can be removed. You may also need to remove a few downstream asserts too.

Reply:
Yeah, you are right. It works. But I'm still confused about the ordering, what exactly must be in post-order?

Reply:
This is tree traversal terminology:

```
fgMorphSmpOp:
   Pre-Order (before the operands were morphed)
   Morphing the operands
   Post-Order (after the operands were morphed)
```
As a general rule, optimizations should be done in post-order, because then we will know facts about the operands we wouldn't have known in pre-order (such as that they are constant, or have side effects).

Reply:
> This is tree traversal terminology:
> 
> ```
> fgMorphSmpOp:
>    Pre-Order (before the operands were morphed)
>    Morphing the operands
>    Post-Order (after the operands were morphed)
> ```
> 
> As a general rule, optimizations should be done in post-order, because then we will know facts about the operands we wouldn't have known in pre-order (such as that they are constant, or have side effects).

Oh, I know about the terminology. Thank you for the explanation :)